front/lib/api/assistant/actions/action_file_helpers.ts

```diff
@@ -0,0 +1,224 @@
+ */
+export async function generateCSVFileAndSnippet(
+  auth: Authenticator,
+  {
+    title,
+    conversationId,
+    results,
+  }: {
+    title: string;
+    conversationId: string;
+    results: Array<CSVRecord>;
+  }
+): Promise<{
+  csvFile: FileResource;
+  csvSnippet: string;
+}> {
+  const workspace = auth.getNonNullableWorkspace();
+  const user = auth.user();
+
+  const csvOutput = await toCsv(results);
+  const csvFile = await FileResource.makeNew({
+    workspaceId: workspace.id,
+    userId: user?.id ?? null,
+    contentType: "text/csv",
+    fileName: title,
+    fileSize: Buffer.byteLength(csvOutput),
+    useCase: "tool_output",
+    useCaseMetadata: {
+      conversationId,
+    },
+  });
+  const csvSnippet = generateCSVSnippet({
+    content: csvOutput,
+    totalRecords: results.length,
+  });
+
+  await processAndStoreFile(auth, { file: csvFile, reqOrString: csvOutput });
+
+  return { csvFile, csvSnippet };
+}
+
+/**
+ * Generate a searchable text file. This type of file is used to store the results of a
+ * tool call coming up from a csv in a way that can be searched.
+ * Save it to the database and return it.
+ */
+export async function generateSearchableFile(
+  auth: Authenticator,
+  {
+    title,
+    conversationId,
+    results,
+    connectorProvider,
+  }: {
+    title: string;
+    conversationId: string;
+    results: Array<CSVRecord>;
+    connectorProvider: ConnectorProvider | null;
+  }
+): Promise<FileResource> {
+  const workspace = auth.getNonNullableWorkspace();
+  const user = auth.user();
+
+  // We loop through the results to generate the section of the searchable file.
+  // The content of the file is the JSON representation of the section.
+  const sections: Array<CoreAPIDataSourceDocumentSection> = [];
+  for (const row of results) {
+    const prefix = getSearchableSectionRowPrefix(connectorProvider, row);
+    const rowContent = JSON.stringify(row);
+    const section: CoreAPIDataSourceDocumentSection = {
+      prefix,
+      content: rowContent,
+      sections: [],
+    };
+    sections.push(section);
+  }
+  const section = {
+    prefix: title,
+    content: null,
+    sections,
+  };
+  const content = JSON.stringify(section);
+
+  const searchableFile = await FileResource.makeNew({
+    workspaceId: workspace.id,
+    userId: user?.id ?? null,
+    contentType: "application/vnd.dust.section-structured",
+    fileName: title,
+    fileSize: Buffer.byteLength(content),
+    useCase: "tool_output",
+    useCaseMetadata: {
+      conversationId,
+    },
+  });
+
+  await processAndStoreFile(auth, {
+    file: searchableFile,
+    reqOrString: content,
+  });
+
+  return searchableFile;
+}
+
+/**
+ * Get the prefix for a row in a searchable section.
+ * This prefix is used to identify the row in the searchable file.
+ * We currently only support Salesforce since it's the only connector for which we can
+ * generate a prefix.
+ */
+function getSearchableSectionRowPrefix(
```

Comment:
This if == salesforce is pretttty isolated, also ConnectorProvider feels a bit off here. Can we pass the function as argument to `generateSearchableFile` and have that logic live in tables_query which is much "closer" to the concept of ConnectorProvider?

Also maybe have a switch with a return null for all providers + assertNever so that we are required to think about it when adding a new provider?

Reply:
I moved it, and switched to switch assertNever pattern. I understand why we want to do that and at the same time this adds even more boilerplate for the next connector, while probably not relevant for it :/